### PR TITLE
fix: invalid request IP

### PR DESCRIPTION
**Pull request by Chris6077**
Checklist:

- [X] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [X] My pull request has a descriptive title (not a vague title like `Update index.md`)

Closes https://github.com/freeCodeCamp/freeCodeCamp/issues/37700

req.connection.remoteAddress delivers the same IP for different devices and connections

### CODE REVIEW

**Comment by scissorsneedfoodtoo** [comment]
Thanks for your patience @Chris6077, and thanks again for submitting this fix.

I tried running this locally and the server crashed. I think it's because `req.get('x-forwarded-for')` returned `undefined`, so there was nothing for the `.split()` method to work with.

I'll leave a review with a suggestion to fix this.

File: stock-price-checker/routes/api.js [hunk 1]
```diff
@@ -19,7 +19,7 @@ module.exports = function (app) {
     .get(function (req, res){
       var stock = req.query.stock;
       var like = req.query.like || false;
-      var reqIP = req.connection.remoteAddress;
+      var reqIP = req.get('x-forwarded-for').split(",")[0] || req.ip.split(",")[0];
```
Comment (scissorsneedfoodtoo) [inline stock-price-checker/routes/api.js:22]: Optional chaining seems to fix this while running locally:

```suggestion
      var reqIP = req.get('x-forwarded-for')?.split(",")[0] || req.ip?.split(",")[0];
```

**Review by scissorsneedfoodtoo — CHANGES_REQUESTED** [review]

**Review by raisedadead — APPROVED** [review]
Thanks @Chris6077 & @scissorsneedfoodtoo 

Hey @freeCodeCamp/dev-team please approve and merge. Let me know when ready, to deploy.

**Review by moT01 — COMMENTED** [review]
I was getting an error with this...
```
  var reqIP = req.get('x-forwarded-for')?.split(",")[0] || req.ip?.split(",")[0];
                                         ^
SyntaxError: Unexpected token '.'
```
Found the problem I think - I was using node v12.16.1 - switching to 14.4 worked. I took a quick look at the server, seems to be using 12.21.0...
```
freecodecamp@demo-projects-vm:~$ node --version
v12.21.0
```
I switched to that version locally, got the error again. So we should either update node on the server? Or switching to something like this in the code seemed to work:
```
var requestIP = req.get('x-forwarded-for') || req.ip;
const reqIP = requestIP.split(",")[0]
```

**Comment by scissorsneedfoodtoo** [comment]
Oof, thanks for catching that @moT01. I didn't think to check the version of Node running on the server.

I'm fine with switching to the code you suggested. I'm sure it won't be a problem to switch to Node 14, but if we do, we should probably give the other projects a once over, just in case there's some other incompatibility.

Your changes would be a bit quicker and easier to implement.

**Comment by scissorsneedfoodtoo** [comment]
Thanks again @moT01. Tested your suggestions locally and everything LGTM.

This may be out of scope for this PR, but as mentioned in https://github.com/freeCodeCamp/freeCodeCamp/issues/37700, it may be good to anonymize IP addresses as they come in, before they're saved to the db.

I'll make a suggestion above in case we want to do that here for the demo project. Though if we want to help learners stay compliant with GDPR and other regulations, we'll need to add a note in the existing description to either hash IP addresses, or truncate parts of it.

File: stock-price-checker/routes/api.js [hunk 2]
```diff
@@ -19,7 +19,8 @@ module.exports = function (app) {
     .get(function (req, res){
       var stock = req.query.stock;
       var like = req.query.like || false;
-      var reqIP = req.connection.remoteAddress;
+      var requestIP = req.get('x-forwarded-for') || req.ip;
+      var reqIP = requestIP.split(",")[0];
```
Comment (scissorsneedfoodtoo) [inline stock-price-checker/routes/api.js:23]: Tested some regex from [this SO post](https://stackoverflow.com/questions/48767382/anonymize-ipv4-and-ipv6-addresses-with-php-preg-replace) and it seems to work for both IPv4 and IPv6 addresses:

```suggestion
      var requestIP = req.get('x-forwarded-for') || req.ip;
      var reqIP = requestIP.split(",")[0];
      reqIP = reqIP.replace(/\.\d*$/, '.0'); // ipv4
      reqIP = reqIP.replace(/[\da-f]*:[\da-f]*$/, '0:0'); // ipv6
```

There's probably a smarter way to do this, but the code above works in a pinch.

Comment (moT01) [reply stock-price-checker/routes/api.js:23]: I think we would need to ignore the case on those v6 versions

**Review by moT01 — APPROVED** [review]
LGTM 🎉 Want to give this one last look @scissorsneedfoodtoo

**Comment by scissorsneedfoodtoo** [comment]
Awesome, thanks for taking another look @moT01. I checked again locally and everything LGTM.

Thanks again for your help with this @Chris6077.

**Pull request merged**